frontend/document/credocument.lua

```diff
@@ -283,10 +288,150 @@ function CreDocument:updateColorRendering()
     end
 end
 
+function CreDocument:setHideNonlinearFlows(hide_nonlinear_flows)
+    if hide_nonlinear_flows ~= self.hide_nonlinear_flows then
+        self.hide_nonlinear_flows = hide_nonlinear_flows
+        self._document:setIntProperty("crengine.doc.nonlinear.pagebreak.force", self.hide_nonlinear_flows and 1 or 0)
+        self:cacheFlows()
+    end
+end
+
 function CreDocument:getPageCount()
     return self._document:getPages()
 end
 
+-- Whether the document has any non-linear flow to care about
+function CreDocument:hasNonLinearFlows()
+    return self._document:hasNonLinearFlows()
+end
+
+-- Whether non-linear flows (if any) will be hidden
+function CreDocument:hasHiddenFlows()
+    return self.flows[1] ~= nil
+end
+
+-- Get the next/prev page number, skipping non-linear flows,
+-- i.e. the next/prev page that is either in the current
+-- flow or in the linear flow (flow 0)
+-- If "page" is 0, these give the initial and final linear pages
+function CreDocument:getNextPage(page)
+    if self:hasHiddenFlows() then
+        if page < 0 then return 0 end
+        local last_linear_page = self:getLastLinearPage()
+        local flow = page > 0 and self:getPageFlow(page) or 0
+        for test_page=page+1, self:getPageCount() do
+            local test_page_flow = self:getPageFlow(test_page)
+            if test_page_flow == flow or test_page_flow == 0 then
+                return test_page
+            end
+            if test_page_flow ~= flow and test_page > last_linear_page then
+                break
```

Comment:
Can't you have/use a getLastPageInFlow(flow) when flow > 0 (like you have in the next function) ?
As when a flow is done, you can't meet it again - so it's no use checking up to the last linear page. No?

Reply:
You can't meet it again, but you will meet flow 0 eventually (unless `test_page > last_linear_page`), and that will be the "next page". I could limit the loop to `math.max(self.getLastPageInFlow(flow), self.getLastLinearPage())`.

Reply:
Or I misunderstood this function :) 
`-- Get the next/prev page number, skipping non-linear flows,`
I thought you could use it to navigate pages inside a non-linear flow - and that if you're in flow No7 out of 19, you don't need to test pages until the end of the document.

Reply:
You skip non-linear flows, but linear pages may be interspersed between them, and we want to stop there.

Reply:
But you really don't use CreDocument:getNextPage(p) to page forward inside non-linear flows (and expecting to get the next page inside this same non-linear flow)?
OK, I guess you do (so, this comment is lying :) and you'll stop looping when you meet back a page from flow 0.

Reply:
There could be another performance improvement: when a different non-linear flow is found, instead of trying pages one by one, I could in principle skip all *n* pages in the flow in one go. But I somehow liked the simplicity of just paging through, it makes fewer assumptions about how the flows behave (for instance, non-linear flows should not be interleaved, but the one-by-one method would work even if that was the case).